SI-6640 Better reporting of deficient classpaths.

retronym opened the pull request:
In a55788e, StubSymbols were introduced to fail-slow when
the classpath was deficient. This allowed compilation to
succeed in cases when one didn't actually use the part of
class A which referred to some missing class B.

But a few problems were introduced.

Firstly, when the deferred error eventually happened, it was
signalled with abort(msg), rather than through a thrown
MissingRequirementError. The latter is desirable, as it doesn't
lead to printing a stack trace.

Second, the actual error message changed, and no longer
included the name of the class file that refers to the missing
class.

Finally, it seems that we can end up with a stub term symbol
in a situation where a class symbol is desired. An assertion
in the constructor of ThisType throws trips when calling .isClass,
before the useful error message from StubSymbol can be emitted.

This commit addresses these points, and rewords the error
a little to be more accessible. The last point is the most fragile
in this arrangement, there might be some whack-a-mole
required to similar assertions. I don't see a clean solution
for this, but am open to suggestions.

After the change:

```
[info] Set current project to default-821d14 (in build file:/Users/jason/code/scratch1/)
> compile

[info] Compiling 1 Scala source to /Users/jason/code/scratch1/target/scala-2.10/classes...
[error]
[error]      while compiling: /Users/jason/code/scratch1/test.scala
[error]         during phase: typer
[error]      library version: version 2.10.0-RC2
[error]     compiler version: version 2.10.0-RC2
...
[error]   last tree to typer: Ident(SwingWorker)
[error]               symbol: <none> (flags: )
[error]    symbol definition: <none>
[error]        symbol owners:
[error]       context owners: object Test -> package <empty>
 ...
[error] uncaught exception during compilation: java.lang.AssertionError
[trace] Stack trace suppressed: run last compile:compile for the full output.
[error] (compile:compile) java.lang.AssertionError: assertion failed: value actors
[error] Total time: 2 s, completed Nov 10, 2012 3:18:34 PM
>
> set scalaHome := Some(file("/Users/jason/code/scala/build/pack"))
[info] Defining *:scala-home
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to default-821d14 (in build file:/Users/jason/code/scratch1/)
^[compile
[info] Compiling 1 Scala source to /Users/jason/code/scratch1/target/scala-2.10/classes...
[error] /Users/jason/code/scratch1/test.scala:4: A signature in SwingWorker.class refers to term actors in package scala which is missing from the classpath.
[error] object Test extends SwingWorker
[error]                     ^
[error] one error found
[error] (compile:compile) Compilation failed
[error] Total time: 2 s, completed Nov 10, 2012 3:18:45 PM
```

Review by @paulp @jsuereth

retronym committed:
SI-6640 Better reporting of deficient classpaths.

In a55788e, StubSymbols were introduced to fail-slow when
the classpath was deficient. This allowed compilation to
succeed in cases when one didn't actually use the part of
class A which referred to some missing class B.

But a few problems were introduced.

Firstly, when the deferred error eventually happened, it was
signalled with abort(msg), rather than through a thrown
MissingRequirementError. The latter is desirable, as it doesn't
lead to printing a stack trace.

Second, the actual error message changed, and no longer
included the name of the class file that refers to the missing
class.

Finally, it seems that we can end up with a stub term symbol
in a situation where a class symbol is desired. An assertion
in the constructor of ThisType throws trips when calling .isClass,
before the useful error message from StubSymbol can be emitted.

This commit addresses these points, and rewords the error
a little to be more accessible. The last point is the most fragile
in this arrangement, there might be some whack-a-mole
required to find other places that also need this.
I don't see a clean solution for this, but am open to suggestions.

We should really build a facility in partest to delete
specified classfiles between groups in separate compilation
tests, in order to have tests for this. I'll work on that as a followup.

For now, here's the result of my manual testing:

[info] Set current project to default-821d14 (in build file:/Users/jason/code/scratch1/)
> compile

[info] Compiling 1 Scala source to /Users/jason/code/scratch1/target/scala-2.10/classes...
[error]
[error]      while compiling: /Users/jason/code/scratch1/test.scala
[error]         during phase: typer
[error]      library version: version 2.10.0-RC2
[error]     compiler version: version 2.10.0-RC2
...
[error]   last tree to typer: Ident(SwingWorker)
[error]               symbol: <none> (flags: )
[error]    symbol definition: <none>
[error]        symbol owners:
[error]       context owners: object Test -> package <empty>
 ...
[error] uncaught exception during compilation: java.lang.AssertionError
[trace] Stack trace suppressed: run last compile:compile for the full output.
[error] (compile:compile) java.lang.AssertionError: assertion failed: value actors
[error] Total time: 2 s, completed Nov 10, 2012 3:18:34 PM
>
> set scalaHome := Some(file("/Users/jason/code/scala/build/pack"))
[info] Defining *:scala-home
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to default-821d14 (in build file:/Users/jason/code/scratch1/)
^[compile
[info] Compiling 1 Scala source to /Users/jason/code/scratch1/target/scala-2.10/classes...
[error] /Users/jason/code/scratch1/test.scala:4: A signature in SwingWorker.class refers to term actors in package scala which is missing from the classpath.
[error] object Test extends SwingWorker
[error]                     ^
[error] one error found
[error] (compile:compile) Compilation failed
[error] Total time: 2 s, completed Nov 10, 2012 3:18:45 PM

xeno-by commented:
Nice error message!

xeno-by commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -3124,12 +3121,10 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
-    locally {
-      debugwarn("creating stub symbol for " + stubWarning)
-    }
+    debugwarn("creating stub symbol to defer error: " + missingMessage)
```
I wonder why Paul wrote `locally` here.

retronym replied:
I'd wager that as he was working on this he few vals before that line that he didn't want retained as fields.

paulp replied:
I don't like the way constructor bodies blend seamlessly with member definitions and intersperse arbitrarily; sometimes I use locally only to bring emphasis to the fact that a constructor exists.

xeno-by commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1385,7 +1385,7 @@ trait Types extends api.Types { self: SymbolTable =>
    */
   abstract case class ThisType(sym: Symbol) extends SingletonType with ThisTypeApi {
-    assert(sym.isClass, sym)
+    assert(sym.isClass, {sym.info; sym}) // call .info to allow StubSymbols to reveal what's missing from the classpath
```
Mmmm. This asks for cyclic reference errors. Are you sure?

retronym replied:
Maybe `sym.forceStubFailure()` would be clearer/safer.

xeno-by commented on src/reflect/scala/reflect/internal/pickling/UnPickler.scala:
```diff
@@ -233,7 +233,8 @@ abstract class UnPickler /*extends scala.reflect.generic.UnPickler*/ {
               adjust(mirrorThatLoaded(owner).missingHook(owner, name)) orElse {
                 // (5) Create a stub symbol to defer hard failure a little longer.
-                owner.newStubSymbol(name)
+                val missingMessage = s"A signature in $filename refers to ${name.longString} in ${owner.fullLocationString} which is missing from the classpath."
```
It looks like `missingMessage` is created here just because `filename` is N/A at the error throwing point.

If that's true, then probably `symbol.associatedFile` would help to evict this nuisance.

retronym replied:
That was used in the previous error message, but didn't show anything. 

We don't want the file of this symbol, but the file of the referring class.

xeno-by commented on src/reflect/scala/reflect/internal/pickling/UnPickler.scala:
```diff
@@ -827,11 +828,6 @@ abstract class UnPickler /*extends scala.reflect.generic.UnPickler*/ {
-      mirrorThatLoaded(owner).missingHook(owner, name) orElse MissingRequirementError.signal(
-        s"bad reference while unpickling $filename: ${name.longString} not found in ${owner.tpe.widen}"
-      )
-
```
Why has this method been removed? upd. Interesting... Apparently it wasn't used at all!

retronym replied:
It was used before Paul's patch.

retronym commented:
I'm also thinking I've narrowed the error message too far. I suppose you could get the same error when using an incorrect version of a classfile, not just when it is missing completely.

retronym committed:
Refine the message and triggering of MissingRequirementError.

 - To force a failure of the stub, call a new method `failIfStub`
   rather than `info`.
 - Offer a broader range of potential root causes in the
   error message.

retronym commented:
@xeno-by I've tried to make the trip-wire for this error more clear, and I've also expanded the error message with other possibilities.

gkossakowski commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1385,7 +1385,11 @@ trait Types extends api.Types { self: SymbolTable =>
   /** A class for this-types of the form <sym>.this.type
    */
   abstract case class ThisType(sym: Symbol) extends SingletonType with ThisTypeApi {
+    // SI-6440 allow StubSymbols to reveal what's missing from the classpath
```
You meant SI-6640 here, right?

gkossakowski commented:
How do we know that call to `sym.failIfStub()` is only needed in `ThisType`?

retronym commented:
Comment fixed.

We don't know if it's the only place; that's the "whack-a-mole" i referred to in the commit comment. The stub symbols intercept calls to `info` which catches the vast majority of places; I looked around for other calls to `isClass` assertions in the vicinity and couldn't find others.

To flush more out, we could take a big codebase with two modules (lets call them "lib" an "app"), compile it all, delete some classfiles at random from "lib", and recompile "app". That should never crash.

But the crash only comes in place of an otherwise failed compilation, so I think we can take some risk here.

retronym commented:
:cat: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/662/console

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1614/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/910/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1614/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/910/

retronym committed:
SI-6440 Address regressions around MissingRequirementError

 Go back to using globalError to report when a stub's info is referenced,
 and only throw the MissingRequirementError when compilation really
 must abort due to having a StubTermSymbol in a place where a
 StubClassSymbol would have been a better choice.

 This situation arises when an entire package is missing from the
 classpath, as was the case in the reported bug.

 Adds `StoreReporterDirectTest`, which buffers messages issued
 during compilation for more structured interrogation. Use this
 in two test for manifests -- these tests were using a crude means
 of grepping compiler console output to focus on the relevant output,
 but this approach was insufficient with the new multi-line error
 message emitted as part of this change.

 Also used that base test class to add two new tests: one for
 the reported error (package missing), and another for a simpler
 error (class missing). The latter test shows how stub symbols
 allow code to compile if it doesn't the subset of signatures
 in some type that refer to a missing class.

 Gave the INFO/WARNING/ERROR members of Reporter sensible
 toString implementations; they inherit from Enumeration#Value
 in an unusual manner (why?) that means the built in toString of
 Enumeration printed `Severity@0`.

retronym commented:
This was more subtle than I originally thought. I've addressed the failing tests and added some new ones in the latest commit.

Review by @xeno-by @paulp

paulp commented on test/files/run/t6440.check:
```diff
@@ -1 +1,4 @@
+pos: source-newSource1,line-9,offset=109 bad symbolic reference. A signature in U.class refers to term pack1
+in package <root> which is not available.
+It may be completely missing from the current classpath, or the version on
+the classpath might be incompatible with the version used when compiling U.class. ERROR
```
Ending with ERROR has a bit of a DOS feel.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/924/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/924/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1628/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1628/

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1385,7 +1385,12 @@ trait Types extends api.Types { self: SymbolTable =>
   /** A class for this-types of the form <sym>.this.type
    */
   abstract case class ThisType(sym: Symbol) extends SingletonType with ThisTypeApi {
-    assert(sym.isClass, sym)
```
one way to whack the mole more reliably and get more encapsulation to boot:
leave this intact and provide failing implementations of all Symbol methods in the StubSymbol subclasses

I'm figuring out whether this is doable. There's also risk with missing an override of course.

retronym replied:
The other stub methods return `NoType` / `NoSymbol` etc., but what do we do here?

We need to know what the callers expectation is.

``` scala
def checkIsClass(expectation: Boolean) = if (expectation != isClass) abort(...)
```

So... tricky.

adriaanm replied:
I imagined isClass on a StubSymbol would throw the MissingRequirementException

retronym replied:
Yeah, that would nicely encapsulate the trigger, but it would (potentially) blow up more often that it does now. Because we don't have tests for all the variations, I erred on the side of "crash rather than fail with a nice error" rather than "fail with a nice error rather than soldier on."

While I could be persuaded to the other side, I just wanted to point out we can't encapsulate without changing behaviour.

For sure we should revisit this. But right now, the most valuable input for the PR would be additional test cases (even ideas for them, I'm happy to implement them.)

gkossakowski commented:
@retronym: I think your approach of using DirectTest and capturing reported errors make perfect sense. I have an impression that you are looking for better alternative. What's the problem with this approach?

retronym commented:
@gkossakowski I `push -f`-ed, but forgot to update the PR description. The `DirectTest` approach is workable, although quite slow.

adriaanm commented:
LGTM, thanks @retronym -- it's great this now comes with tests
